Automap: Transparency

kphoenix137 opened the pull request:
Transparent automap.

Usage: Press Tab or Map button to open automap (opaque), press again to make automap transparent, press again to close automap

The reason that the automap looks so patchy with transparency is because the line drawing functions are horrible about not drawing over each other, causing doubling of drawn pixels, which makes them less transparent. I think it looks kinda cool though, as the lines look more like the golden borders in the game.

![image](https://github.com/diasurgical/devilutionX/assets/68359262/498b9c41-60d6-424e-99fb-553118fc4f21)

AJenbo commented:
Needs a rebase

FitzRoyX commented:
I only think it should be a setting like auto gold pickup, or maybe assign T to it. People are used to automap being a quick on/off toggle, and this would force you to cycle pas it if you didn't want it.

AJenbo commented:
Space will still do that

FitzRoyX commented:
> Space will still do that

Space closes out panels too, though. I'm not sure how many are using that as their automap disable key.

glebm commented:
How about introducing a separate hotkey for cycling the map type?

E.g. "m": opaque -> transparent -> minimap -> off

On the gamepad, the regular toggle map would remain the same (`LEFTSTICK`), and the `START + DOWN` would now cycle the map type.

This way we won't need to add any new options.

glebm commented:
@kphoenix137 I've implemented the proposal above in #6613

AJenbo commented:
I merged https://github.com/diasurgical/devilutionX/pull/6613, please use it for the type changing :)

kphoenix137 commented:
@AJenbo should be ready. Also to go over some of the changes in this PR:

I added `SetMapPixel()` as a way to decide which pixel drawing function should be used based on if the map is transparent or not. I also added a function that draws transparent pixels, since we previously didn't have this. It's not perfect though, because transparent pixels on the automap are often drawn over other transparent pixels. I think it looks fine but maybe not to others. If avoiding all overdraw is a requirement, I will need help.

I modified glebm's cycling implementation a bit. I didn't care for the fact that the cycling key was able to turn on/off the entire automap. I prefer a functionality that loops it, such as Opaque - > Transparent - > Minimap - > Opaque , etc. I added a small hack to skip the minimap. I'll remove this for the minimap PR.

For drawing the player arrow on the automap, I had to add 2 new functions, since I couldn't modify the existing line drawing functions without affecting any other calls to that function outside the automap. I'm hopefully I did it in a clean way, but maybe there's a better way to handle it?

glebm commented:
> I prefer a functionality that loops it, such as Opaque - > Transparent - > Minimap - > Opaque

That won't work for devices with limited number of buttons, they're no longer able to turn automap off with this change

kphoenix137 commented:
> > I prefer a functionality that loops it, such as Opaque - > Transparent - > Minimap - > Opaque
> 
> That won't work for devices with limited number of buttons, they're no longer able to turn automap off with this change

Good point, I will add your functionality back in.
